### PR TITLE
Fix issues with props

### PR DESCRIPTION
**Pull request by armano2**
With those changes, its going to be easier to add support for classes in feature 

* `vue/prop-name-casing`: is working now with array props [literals]
* `vue/prop-name-casing`: reports all errors if there are non Literal keys in it
* `vue/require-default-prop`: allow to use shorthand, properties and function calls
* `vue/require-prop-types`: reports names for types different than literals
* `vue/require-prop-types`: fix false positive error in when is set to empty array
* `vue/require-prop-types`: will return now errors about each prop from ArrayExpression
* add new `getComponentProps` helper to easily deal with props 
* fix spread operator issues across all `prop` related rules

fix issue: #596 
fix issue #595

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 .idea
+*.iml
```
Comment (michalsnik) [inline .gitignore]: What is this file?

Comment (armano2) [reply .gitignore]: https://www.jetbrains.org/intellij/sdk/docs/basics/project_structure.html
> Module Library: the library classes are visible only in this module and the library information is recorded in the module’s .iml file.

**Comment by armano2** [comment]
@michalsnik all requested changes are applied, 
additionally i made interface of `getComponentProps` more like `getComputedProperties`

File: lib/rules/require-prop-type-constructor.js [hunk 2]
```diff
@@ -70,31 +70,23 @@ module.exports = {
...
 
     return utils.executeOnVueComponent(context, (obj) => {
-      const node = obj.properties.find(p =>
-        p.type === 'Property' &&
-        p.key.type === 'Identifier' &&
-        p.key.name === 'props' &&
-        p.value.type === 'ObjectExpression'
-      )
+      const props = utils.getComponentProps(obj)
+        .filter(cp => cp.key && cp.value)
```
Comment (michalsnik) [inline lib/rules/require-prop-type-constructor.js]: `cp` should rather be named `prop` here

File: lib/rules/require-prop-type-constructor.js [hunk 3]
```diff
@@ -70,31 +70,23 @@ module.exports = {
...
 
-      if (!node) return
+      for (const p of props) {
```
Comment (michalsnik) [inline lib/rules/require-prop-type-constructor.js]: Can we also use `prop` here instead of `p`?

File: lib/rules/require-prop-type-constructor.js [hunk 4]
```diff
@@ -70,31 +70,23 @@ module.exports = {
...
+        if (isForbiddenType(p.value) || p.value.type === 'ArrayExpression') {
+          checkPropertyNode(p.key, p.value)
+        } else if (p.value.type === 'ObjectExpression') {
+          const typeProperty = p.value.properties.find(prop =>
```
Comment (michalsnik) [inline lib/rules/require-prop-type-constructor.js]: And here we can use `p` or `property` then

File: lib/rules/require-valid-default-prop.js [hunk 5]
```diff
@@ -88,20 +88,11 @@ module.exports = {
...
-        p.value.type === 'ObjectExpression'
-      )
-      if (!props) return
-
-      const properties = props.value.properties.filter(p =>
-        isPropertyIdentifier(p) &&
-        utils.unwrapTypes(p.value).type === 'ObjectExpression'
-      )
+      const props = utils.getComponentProps(obj)
+        .filter(cp => cp.key && cp.value && cp.value.type === 'ObjectExpression')
```
Comment (michalsnik) [inline lib/rules/require-valid-default-prop.js]: `cp` -> `prop`

File: lib/utils/index.js [hunk 6]
```diff
@@ -365,9 +365,46 @@ module.exports = {
...
+
+    if (!propsNode) {
+      return []
+    }
+
+    let props
+
+    if (propsNode.value.type === 'ObjectExpression') {
+      props = propsNode.value.properties
+        .filter(cp => cp.type === 'Property')
```
Comment (michalsnik) [inline lib/utils/index.js]: `cp` -> `prop`, and everywhere else too, expect places where it stands for `computed property` :)

File: tests/lib/rules/require-prop-types.js [hunk 7]
```diff
@@ -149,25 +167,43 @@ ruleTester.run('require-prop-types', rule, {
...
+        message: 'Prop "foo" should define at least its type.',
+        line: 3
+      }, {
+        message: 'Prop "bar" should define at least its type.',
+        line: 3
+      }, {
+        message: 'Prop "baz" should define at least its type.',
+        line: 3
+      }, {
+        message: 'Prop "***" should define at least its type.',
```
Comment (michalsnik) [inline tests/lib/rules/require-prop-types.js]: Perhaps we should call it `Unknown prop`? I won't be ideal though. I think we can do better - perhaps in another PR and only report this once if the props is an Array. Now we'll always get as many errors as we have props in the array anyway.

**Comment by armano2** [comment]
@michalsnik requested changes applied

**Pull request merged**